fdbserver/VersionedBTree.actor.cpp

```diff
@@ -2931,46 +3012,54 @@ class DWALPager final : public IPager2 {
 		debug_printf("DWALPager(%s) op=readPhysicalStart %s ptr=%p\n",
 		             self->filename.c_str(),
 		             toString(pageIDs).c_str(),
-		             page->begin());
+		             page->rawData());
 
-		++g_redwoodMetrics.metric.pagerDiskRead;
 		// TODO:  Could a dispatched read try to write to page after it has been destroyed if this actor is cancelled?
 		state int blockSize = self->physicalPageSize;
-		state uint8_t* data = page->mutate();
 		std::vector<Future<int>> reads;
 		for (int i = 0; i < pageIDs.size(); ++i) {
-			reads.push_back(readPhysicalPage_impl(self, data, blockSize, ((int64_t)pageIDs[i]) * blockSize, priority));
-			data += blockSize;
+			reads.push_back(readPhysicalBlock(
+			    self, page->rawData() + (i * blockSize), blockSize, ((int64_t)pageIDs[i]) * blockSize, priority));
 		}
 		// wait for all the parallel read futures
 		wait(waitForAll(reads));
 
 		debug_printf("DWALPager(%s) op=readPhysicalComplete %s ptr=%p bytes=%d\n",
 		             self->filename.c_str(),
 		             toString(pageIDs).c_str(),
-		             page->begin(),
+		             page->rawData(),
 		             pageIDs.size() * blockSize);
-		// Header reads are checked explicitly during recovery
-		if (!page->verifyChecksum(pageIDs.front())) {
-			debug_printf("DWALPager(%s) checksum failed for %s\n", self->filename.c_str(), toString(pageIDs).c_str());
-			Error e = checksum_failed();
-			TraceEvent(SevError, "RedwoodChecksumFailed")
+
+		try {
+			page->postRead1(pageIDs.front());
+			if (self->pEncryptionKeyProvider != nullptr && page->isEncrypted()) {
+				Key secret = wait(self->pEncryptionKeyProvider->getSecretByKeyID(page->getKeyID()));
+				page->setSecret(secret);
+			}
+			page->postRead2(pageIDs.front());
+		} catch (Error& e) {
+			// For header pages, error is a warning because recovery may still be possible
+			TraceEvent(SevError, "RedwoodPageError")
```

Comment:
a nit: SevError -> SevWarn?

Reply:
Not SevWarn, possibly SevWarnAlways but if we hit this error then either a software bug has occurred or some hardware error has occurred that corrupted data.  When hardware returns an explicit error we log it as io_error with SevError, so that seems appropriate here.